### PR TITLE
#362 work package edit form

### LINKED ISSUE #362: Projects - Work Package Edit Form
## Description

A new component needs to be created so that project managers can edit a work package. This component will be a form that will look similar to the work package details page, except that all the fields within within the work package will be editable text boxes rather than view-only details about the work package. 

## Background
The idea is that when project managers click on the "edit" button (which will be completed in another ticket), this form will be populated with the current information of the work package, so that project managers can directly edit and save the details. This ticket is to make the form that the project managers will see when they click the "edit" button. Do not worry about populating the text boxes with the current values, as that will be completed in another ticket. The focus of this ticket is to only create the front-end visual.

## Proposed Solution
\*\***For all sections listed below, the headers for each section (Work Package Details, Dependencies, etc.) should not be editable**\*\*

### Work Package Details 
- [x] The labels for each field will remain exactly the same and in the same location. Only the details portion for each field will change and will become input text boxes. 
- [x] The "Start Date" field should become a date picker.
- [x] The "End Date" field should be read only, since it is calculated by the start date and duration.
- [x] The percent symbol for the "Progress" and "Expected Progress" fields should be a read-only field. The user will only be changing the numbers. See mocks for clarification.
- [x] The "weeks" unit for the "Duration" field should be a read-only field. The user will only be changing the numbers. See mocks for clarification.
- [x] The "status" field on the top right corner of this section should be a dropdown list with the following three selections: "Inactive", "Active", "Complete".

### Dependencies
- [x] Each existing dependency will become a text box with a delete button on the upper left hand corner of each text box. 
- [x] At the end of the list, there should be a blank text box with an add button on the right hand side of the text box. This text box and button should always be present, whether or not there are any list items present.

### Expected Activities
- [x] Each existing expected activity will become a text box with a delete button for each line item located on the right hand side of the screen.
- [x] At the end of the list, there should be a blank text box with an add button on the right hand side of the text box. This text box and button should always be present, whether or not there are any list items present.

### Deliverables
- [x] Each existing deliverable will become a text box with a delete button for each line item located on the right hand side of the screen.
- [x] At the end of the list, there should be a blank text box with an add button on the right hand side of the text box. This text box and button should always be present, whether or not there are any list items present.

### Changes
- [x] View only section. The section should not be editable. It should still be a part of the form, but should remain a view-only section.

### General 
- [x] There should be a sticky footer with two buttons: "Cancel" and "Save". The footer will be slightly transparent.

## Mocks
https://www.figma.com/file/qmXuySSwfiD942HwmSSEMe/NERPM-Desktop?node-id=392%3A225

### PR DESCRIPTION
**Pull request by z-roth**
### What's New:
**Work Package Buttons** - New buttons that say New Change Request (not functional) and Edit (places page into edit mode)
**Editable Detail Component** - Used to switch a detail (like the ones in the work package details) from display mode to edit mode, with inputs using their previous values as the placeholders. 
**Editable Dependencies** - Dependencies are now displayed as buttons, and can be deleted when in edit mode. An add dependency button appears when in edit mode as well. 
**Editable Description Lists** - The description lists used for Expected Activities and Deliverables can now go into edit mode, allowing each bullet to be deleted or edited, and an add bullet field and button is added to the end.
**Edit Mode Options** - When edit mode is enabled, two buttons appear at the bottom of the screen, allowing the user to save their changes (nonfunctional) or cancel/disable edit mode.

### Closes #362

### CODE REVIEW

File: src/components/projects/wbs-details/work-package-container/dependencies-list/dependency/dependency.module.css [hunk 1]
```diff
@@ -0,0 +1,9 @@
+/*
+ * This file is part of NER's PM Dashboard and licensed under GNU AGPLv3.
+ * See the LICENSE file in the repository root folder for details.
+ */
+
+ .dependencyButton {
+  background-color: #ef4345 !important;
+  border-color: #ef4345 !important;
+}
```
Comment (jamescd18) [inline src/components/projects/wbs-details/work-package-container/dependencies-list/dependency/dependency.module.css:9]:
```suggestion
}

```

File: src/components/projects/wbs-details/work-package-container/edit-mode-options/edit-mode-options.module.css [hunk 2]
```diff
@@ -0,0 +1,10 @@
+.editModeOptionsContainer {
```
Comment (jamescd18) [inline src/components/projects/wbs-details/work-package-container/edit-mode-options/edit-mode-options.module.css:1]: repo license snippet

File: src/components/projects/wbs-details/work-package-container/edit-mode-options/edit-mode-options.test.tsx [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import { render, screen } from '../../../../../test-support/test-utils';
```
Comment (jamescd18) [inline src/components/projects/wbs-details/work-package-container/edit-mode-options/edit-mode-options.test.tsx:1]: repo license snippet

File: src/components/projects/wbs-details/work-package-container/work-package-buttons/work-package-buttons.module.css [hunk 4]
```diff
@@ -0,0 +1,11 @@
+.workPackageActionButtonsContainer {
```
Comment (jamescd18) [inline src/components/projects/wbs-details/work-package-container/work-package-buttons/work-package-buttons.module.css:1]: repo license snippet

File: src/components/projects/wbs-details/work-package-container/work-package-buttons/work-package-buttons.tsx [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import { useContext } from 'react';
+import { Button } from 'react-bootstrap';
+import { EditModeContext, EditModeProps } from '../work-package-container';
+import styles from './work-package-buttons.module.css';
+
+const WorkPackageButtons: React.FC<EditModeProps> = (props) => {
+  const editMode = useContext(EditModeContext);
+  return (
+    <div className={`mx-4 my-3 ${styles.workPackageActionButtonsContainer}`}>
+      <Button className={styles.workPackageActionButton}>New Change Request</Button>
```
Comment (jamescd18) [inline src/components/projects/wbs-details/work-package-container/work-package-buttons/work-package-buttons.tsx:15]: let's either comment this button out, or add functionality to make it redirect to the new change request page

**Comment by z-roth** [comment]
Attached are screenshots of an example work package, with edit mode disabled and enabled
Work Package Page (Edit Mode Enabled)
Work Package Page (Edit Mode Disabled)
.

File: src/components/projects/wbs-details/work-package-container/work-package-details/work-package-details.tsx [hunk 6]
```diff
@@ -22,29 +19,75 @@ interface WorkPackageDetailsProps {
...
+              type="number"
+            />
+          </Col>
+        </Row>
+      </Form>
     </Container>
   );
 
+  const editMode = useContext(EditModeContext);
+  const statuses = ['Inactive', 'Active', 'Complete'];
```
Comment (jamescd18) [inline src/components/projects/wbs-details/work-package-container/work-package-details/work-package-details.tsx:69]: enum already exists in `utils` package, so you should be able to just do `Object.values(...)`
https://github.com/Northeastern-Electric-Racing/PM-Dashboard-v2/blob/dca5d7b78d63c608c9cac301fdd9b27361f35cd4/src/utils/src/types/project-types.ts#L27

File: src/components/shared/bullet-list/bullet-list.tsx [hunk 7]
```diff
@@ -5,21 +5,47 @@
...
     builtList = <ol className={styles.bulletList}>{listPrepared}</ol>;
   }
+  if (editMode) {
+    builtList = <Form>{listPrepared}</Form>;
+  }
   return <PageBlock title={title} headerRight={headerRight} body={builtList} />;
 };
 
-export default BulletList;
+export default BulletList;
```
Comment (jamescd18) [inline src/components/shared/bullet-list/bullet-list.tsx:51]:
```suggestion
export default BulletList;

```

File: src/components/shared/description-list/description-list.tsx [hunk 8]
```diff
@@ -18,7 +19,7 @@ const DescriptionList: React.FC<DescriptionListProps> = ({ title, items }) => {
       title={title}
       headerRight={<></>}
       list={items.map((b) => (
-        <>{b.detail}</>
+        <Fragment>{b.detail}</Fragment>
```
Comment (jamescd18) [inline src/components/shared/description-list/description-list.tsx:22]:
```suggestion
        <>{b.detail}</>
```
shorthand syntax FTW ;p

File: src/components/shared/description-list/description-list.tsx [hunk 9]
```diff
@@ -3,6 +3,7 @@
  * See the LICENSE file in the repository root folder for details.
  */
 
+import { Fragment } from 'react';
```
Comment (jamescd18) [inline src/components/shared/description-list/description-list.tsx:6]:
```suggestion
```

File: src/components/shared/editable-detail/editable-detail.test.tsx [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+      <EditModeContext.Provider value={false}>
+        <EditableDetail
+          title="testTitle"
+          value="testValue"
+          readOnly={false}
+          suffix="testSuffix"
+          type="text"
+        />
+      </EditModeContext.Provider>
+    );
```
Comment (jamescd18) [inline src/components/shared/editable-detail/editable-detail.test.tsx:22]: if this is being repeated a bunch with only 1 or 2 values changing, then let's refactor it out into a separate `renderComponent()` function like here: https://github.com/Northeastern-Electric-Racing/PM-Dashboard-v2/blob/dca5d7b78d63c608c9cac301fdd9b27361f35cd4/src/components/change-requests/change-requests.test.tsx#L49

File: src/components/shared/editable-detail/editable-detail.tsx [hunk 11]
```diff
@@ -0,0 +1,46 @@
+/*
+ * This file is part of NER's PM Dashboard and licensed under GNU AGPLv3.
+ * See the LICENSE file in the repository root folder for details.
+ */
+
+import React, { useContext } from 'react';
```
Comment (jamescd18) [inline src/components/shared/editable-detail/editable-detail.tsx:6]:
```suggestion
import { useContext } from 'react';
```
with react 17, you shouldn't need to import the react itself

**Review by jamescd18 — CHANGES_REQUESTED** [review]
really impressive with how complex this all is!

if you're not easily able to get the save and cancel buttons to sticky and stay at the bottom of the screen while scrolling, then can we have them just as the end after the last box at the bottom of the page after scrolling all the way down?

could you add a field for "change request ID" at the top of the page?

be sure to add the repo license snippet to `src/components/shared/editable-detail/editable-detail.module.css` and some other files

remember to resolve the merge conflicts

**Comment by z-roth** [comment]
> if you're not easily able to get the save and cancel buttons to sticky and stay at the bottom of the screen while scrolling, then can we have them just as the end after the last box at the bottom of the page after scrolling all the way down?

I was able to get them to stick to the bottom of the page - they just look a little weird in the screenshot I took

File: src/components/projects/wbs-details/work-package-container/work-package-buttons/work-package-buttons.tsx (hunk 5 above)
Comment (z-roth) [reply src/components/projects/wbs-details/work-package-container/work-package-buttons/work-package-buttons.tsx:15]: the new change request page isn't on this branch - I'll comment it out for now.

**Review by jamescd18 — CHANGES_REQUESTED** [review]
This is looking really really good! Functionality wise though, there's a couple changes we need for MVP launch:
- [x] an input field somewhere on the page (at the top would be nice) that appears only when in edit mode and is labeled "Change Request ID" and is restricted to number inputs
- [x] the WBS number field should not be an editable field because we do not want to allow users to edit that
- [ ] all of the fields right now have the current value being set into the "placeholder" property of the input field, but we actually don't want that, we want the current value to go in as the value of the input field. you many need to do googling to get this done. the only fields that should have placeholder text are the fields for brand-new content like the new bullet fields


Screen Shot 2022-01-17 at 5 07 26 PM

**Review by jamescd18 — APPROVED** [review]
LGTM

**Pull request merged**